### PR TITLE
Batch annotation changes

### PR DESCRIPTION
**Pull request by mj3cheun**
We are using ideogram with way over 2000 annotations and as result when these annotations are added or modifed there is a very noticeable hit to performance which is to be expected. It was found that the reason for this slowdown is that D3 adds all the annotations to the DOM in a single function call regardless of the number of annotations being added. Therefore if the function call takes 5 seconds for example, then the page will effectively freeze for those 5 seconds until all the annotations are added.

The solution I have attempted is the following. While it is not possible to speed up the time required to add annotations to the DOM, we can break up the processing of the annotations, so that instead of trying to process them all in one go, it will process them in multiple batches. The idea is so that in between batches the browser can do whatever else it needs to do to ensure the user experience is smooth and the page is functioning normally. I did find that there is a performance hit in terms of total time required to load the annotations completely (on my computer with 31148 annotations, ~9 seconds with these changes, ~ 4 seconds without). This is the result of the browser repainting multiple times while the annotations load, in addition to executing whatever else it needs to. However in terms of user experience it is much better since there is 0 freezing or noticeable lag on the page and the user gets feedback on the load progress from watching the annotations load in gradually.

Here is a test run loading > 30k annotations.
load

What are your thoughts on this change? Thanks!

Edit: looking at unit test failures

### CODE REVIEW

**Comment by eweitz** [comment]
Plotting 31,148 annotations as individual SVG elements is not ideal. 

It seems you saw the warning when plotting more than 2,000 annotations, which suggests using another [`annotationsLayout`](https://github.com/eweitz/ideogram/blob/master/api.md#annotationslayout), like [`histogram`](https://eweitz.github.io/ideogram/annotations-histogram) or [`heatmap`](https://eweitz.github.io/ideogram/annotations-heatmap) for such large datasets.

What specific use case do you have for plotting over 30,000 annotations?  How do the histogram or heatmap annotation layouts not satisfy it?

**Comment by mj3cheun** [comment]
I completely agree that 30k annotations on the ideogram is not useful. Apologies if it was not clear but I was simply using 30k annotations as an extreme test to showcase the performance that can be achieved by batching the DOM updates.

We noticed that that speed could potentially be an issue when testing the performance of the app with larger number of annotations, other JS running in the background (React, other visualizations, UI), and with slower computers. For example, if another animation is running concurrently then even a brief blockage of the event loop will result in noticeable lag. This change was intended to give users the flexibility to use larger numbers of annotations if they so choose and remove performance as a determining factor in deciding what type and/or how many annotations should be added to the visualization.

However I understand if this change is not worth making, particularly because it increases the complexity of the code, especially given the limited use case as you point out. Otherwise if you believe this is worth looking into further, I believe it should be possible to ensure that this change does not degrade performance for more common use cases by being more intelligent about when batching is used. For example batching can be disabled under a certain threshold of elements to be added.

**Comment by eweitz** [comment]
I am open to this proposal, and want to help it along.  Below are next steps I think are needed.

## Use case
For the number of annotations where you suspect using batched writes significantly improves performance, how does the default annotation layout visually compare with histogram and heatmap layouts?  

A few screenshots showing that the default layout has better visual quality or offers more insight than histograms or heatmaps for such datasets would verify the use case.

## Performance analysis
The proposal does not offer strictly better performance, as you note.  It increases total rendering time for annotations, making it roughly 2.25x slower.  We need to establish the benefit is greater than that cost.

The performance benefit of batched writes seems to be higher frame rate while writing annotations to DOM.  Batching writes only upon crossing a minimum threshold number of annotations sounds promising!

Let's get more data.  Could you please attach four [saved for Chrome DevTools performance profiles](https://developers.google.com/web/tools/chrome-devtools/evaluate-performance/reference#save) (i.e. files containing JSON and named like `Profile-20190613T080000`) for the following scenarios?

1. Batched writes at threshold
2. Unbatched writes at threshold
3. Batched writes at 2x threshold
4. Unbatched writes at 2x threshold

In summary, if we can both verify the use case for this proposed optimization and show its performance benefits exceed its costs, then I think we can move forward.

**Comment by mj3cheun** [comment]
Further work on this PR is going to be deferred to a later date.

**Comment by eweitz** [comment]
Alright, please reference this PR when work resumes.  I will close it for now.

Upon resuming, it be most economical to address [next steps](https://github.com/eweitz/ideogram/pull/152#issuecomment-501689013) for "Use case" and "Performance analysis" before fixing the [failing tests](https://travis-ci.org/eweitz/ideogram/builds/544514277).
